lib/_stream_readable.js

```diff
@@ -700,11 +702,14 @@ Readable.prototype.pipe = function(dest, pipeOpts) {
       // to get stuck in a permanently paused state if that write
       // also returned false.
       // => Check whether `dest` is still a piping destination.
-      if (((state.pipesCount === 1 && state.pipes === dest) ||
+
+      // To avoid increasing the `state.awaitDrain++` incorrectly when in a sync recursion `dest.write(chunk)` call. Here we check if the dest has already been counted on `state.awaitDrain`
+      if (!state.awaitDrainDestSet.has(dest) && ((state.pipesCount === 1 && state.pipes === dest) ||
```

Comment:
Can you add a test, and make sure that `make test` passes locally? I think at least the linter should fail for these lines.

(You probably also want to either run benchmarks yourself or have us do it, because this is likely to have a non-trivial effect.)